### PR TITLE
Introduce GitHub App authentication for GitHub Scaler

### LINKED ISSUE #4651: Harden GitHub Runner Spec
### Proposal

Use a GitHub App to provide a more secure access to GitHub for the runners as opposed to a PAT token and gain increased api usage. (15k vs 5k / hour)

### Use-Case

More secure runner integration with github using certs as opposed to keys

### Is this a feature you are interested in implementing yourself?

Yes

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by Eldarrin**
Implemented Authentication with GitHub App rather than PAT token

### Checklist

- [x] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

Fixes #4651

Relates to https://github.com/kedacore/keda-docs/pull/1159
Relates to https://github.com/kedacore/test-tools/pull/127

### CODE REVIEW

**Comment by JorTurFer** [comment]
> anything I should put in the docs as a note?

No, it was written indeed, the 1s try I used client secret instead of private key, the second one I used the private key in plain text, and the last one I read the documentation you did, and I saw that I should have set the private key in base64...

The problem has been in the keyboard-chair interface xD

**Comment by Eldarrin** [comment]
don't merge it yet, just thought of something I need to check

**Comment by JorTurFer** [comment]
> don't merge it yet, just thought of something I need to check

sure!

**Comment by Eldarrin** [comment]
Its okay, it only relates to test not the scaler. On the e2e we provide pat in clear and use the e2e to b64 it. for the secret it needs to be b64.

File: pkg/scalers/github_runner_scaler.go [hunk 1]
```diff
@@ -403,15 +416,58 @@ func parseGitHubRunnerMetadata(config *ScalerConfig) (*githubRunnerMetadata, err
...
+		return nil, nil, nil, fmt.Errorf("applicationID, installationID and applicationKey must be given")
+	}
+
+	if meta.applicationID != nil && meta.installationID != nil && meta.applicationKey != nil {
+		httpTrans := kedautil.CreateHTTPTransport(false)
+		_, err := gha.New(httpTrans, *meta.applicationID, *meta.installationID, []byte(*meta.applicationKey))
+		if err != nil {
+			return nil, nil, nil, fmt.Errorf("error creating GitHub App client: %w, \n appID: %d, instID: %d", err, appID, instID)
+		}
+	}
```
Comment (JorTurFer) [inline pkg/scalers/github_runner_scaler.go:468]: Do we need to test this here? I mean, we are also checking it during the client creation

**Review by JorTurFer — COMMENTED** [review]
LGTM, only one small nit inline

File: pkg/scalers/github_runner_scaler.go (hunk 1 above)
Comment (Eldarrin) [reply pkg/scalers/github_runner_scaler.go:468]: let me run another test in the morning, I put it in specifically to fix something just can't remember what

Comment (zroubalik) [reply pkg/scalers/github_runner_scaler.go:468]: Yeah, that's a good question. It is not something critical, but would be nice to know if it is really necessary.

**Review by zroubalik — COMMENTED** [review]
Looking great! Thanks for the contribution.

File: pkg/scalers/github_runner_scaler.go (hunk 1 above)
Comment (Eldarrin) [reply pkg/scalers/github_runner_scaler.go]: I've removed the pre check. I was in there to fix a bug, which I subsequently fixed elsewhere and didn't remove it

**Review by zroubalik — COMMENTED** [review]
@Eldarrin could you please fix the conflict? 🙇‍♂️

**Comment by Eldarrin** [comment]
No idea why that is failing, doesn't look like my code

**Review by zroubalik — APPROVED** [review]
LGTM, thanks!

**Comment by piroinno** [comment]
Will you be cutting a new release soon?

**Pull request merged**